src/offset/local/unix.rs

```diff
@@ -37,35 +37,89 @@ pub(super) fn naive_to_local(d: &NaiveDateTime, local: bool) -> LocalResult<Date
+            None => true,
+            _ => false,
+        };
+        if needs_refresh {
+            *cache.borrow_mut() = Some(Cache::new());
+        }
 
-    if local {
-        // we pass through the year as the year of a local point in time must either be valid in that locale, or
-        // the entire time was skipped in which case we will return LocalResult::None anywa.
-        match info
-            .find_local_time_type_from_local(d.timestamp(), d.year())
-            .expect("unable to select local time type")
-        {
-            LocalResult::None => LocalResult::None,
-            LocalResult::Ambiguous(early, late) => LocalResult::Ambiguous(
-                FixedOffset::east(early.offset()),
-                FixedOffset::east(late.offset()),
+        if local {
+            // we pass through the year as the year of a local point in time must either be valid in that locale, or
+            // the entire time was skipped in which case we will return LocalResult::None anywa.
+            match cache
+                .borrow()
+                .as_ref()
+                .unwrap()
+                .zone
+                .find_local_time_type_from_local(d.timestamp(), d.year())
+                .expect("unable to select local time type")
+            {
+                LocalResult::None => LocalResult::None,
+                LocalResult::Ambiguous(early, late) => LocalResult::Ambiguous(
+                    FixedOffset::east(early.offset()),
+                    FixedOffset::east(late.offset()),
+                ),
+                LocalResult::Single(tt) => LocalResult::Single(FixedOffset::east(tt.offset())),
+            }
+        } else {
+            LocalResult::Single(FixedOffset::east(
+                cache
+                    .borrow()
+                    .as_ref()
+                    .unwrap()
+                    .zone
+                    .find_local_time_type(d.timestamp())
+                    .expect("unable to select local time type")
+                    .offset(),
+            ))
+        }
+    })
+}
+
+thread_local! {
+    static TZINFO: RefCell<Option<Cache>> = RefCell::new(None);
+}
+
+enum Source {
+    Localtime(SystemTime),
+    Environment(String),
+}
+
+impl Source {
+    fn new() -> Source {
+        match env::var("TZ") {
+            Ok(var) => Source::Environment(var),
+            Err(_) => Source::Localtime(
+                fs::symlink_metadata("/etc/localtime")
+                    .expect("localtime should exist")
+                    .modified()
+                    .unwrap(),
             ),
-            LocalResult::Single(tt) => LocalResult::Single(FixedOffset::east(tt.offset())),
         }
-    } else {
-        LocalResult::Single(FixedOffset::east(
-            info.find_local_time_type(d.timestamp())
-                .expect("unable to select local time type")
-                .offset(),
-        ))
+    }
+    fn out_of_date(&self) -> bool {
```

Comment:
Style nit: empty line between methods, please.

This still feels like an expensive thing to do on every invocation to `offset()`. Could we amortize the cost somehow?

Reply:
Some thoughts here:
* for the env var variant, this should be very unlikely to change, given the issue surrounding `set_var`, so we could just never alter the cache if the `TZ` env var is set
* In the case it is not set, it could be confusing if there was a lag between symlinking a different timezone and then seeing the change happen - perhaps we could have some configurable behavior with features (eg `never-invalidate-cache`, but as a default always check the `mtime`? If we have a feature it would have to use the previous code with the `Once` rather than thread local storage, to ensure that users don't end up with a different timezone cached on each thread!
* This might interact unexpectedly with asynchronous code (by blocking longer than expected), perhaps adding a new API, say, `LocalCached` where the struct that implements `TimeZone` actually stores the parsed tzinfo file, then users can choose if/when to invalidate the cache

Reply:
1s seems like a reasonable choice in that, it would perhaps be imperceptible to a user manually changing the symlink and then looking at application output. I guess the caching time could be included in the documentation somewhere, such that it is clear to users.

Reply:
Okay, seems good.